### PR TITLE
Completer prefix splitting

### PR DESCRIPTION
**Pull request by gforsyth**
A couple of fixes for better completion support.  Should address #1708 

@laerus, want to take this for a spin?

### CODE REVIEW

**Comment by laerus** [comment]
i'm getting this 

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/interface.py", line 83n
    completions = list(buffer.completer.get_completions(document, complete_event))
  File "/home/laerus/projects/xonsh/xonsh/ptk/completer.py", line 38, in get_completions
    self.ctx)
  File "/home/laerus/projects/xonsh/xonsh/completer.py", line 36, in complete
    out = func(prefix, line, begidx, endidx, ctx)
  File "/home/laerus/projects/xonsh/xonsh/completers/base.py", line 21, in complete_base
    python_comps = complete_python(prefix, line, start, end, ctx)
  File "/home/laerus/projects/xonsh/xonsh/completers/python.py", line 52, in complete_python
    prefix = re.split('\(|=|{|[', prefix)[-1]
  File "/usr/lib/python3.5/re.py", line 203, in split
    return _compile(pattern, flags).split(string, maxsplit)
  File "/usr/lib/python3.5/re.py", line 293, in _compile
    p = sre_compile.compile(pattern, flags)
  File "/usr/lib/python3.5/sre_compile.py", line 536, in compile
    p = sre_parse.parse(p, flags)
  File "/usr/lib/python3.5/sre_parse.py", line 829, in parse
    p = _parse_sub(source, pattern, 0)
  File "/usr/lib/python3.5/sre_parse.py", line 437, in _parse_sub
    itemsappend(_parse(source, state))
  File "/usr/lib/python3.5/sre_parse.py", line 545, in _parse
    source.tell() - here)
sre_constants.error: unterminated character set at position 7
```

File: xonsh/completers/python.py [hunk 1]
```diff
@@ -47,6 +47,12 @@ def complete_python(prefix, line, start, end, ctx):
             rtn |= attr_complete(prefix, ctx, filt)
         rtn |= {s for s in ctx if filt(s, prefix)}
     rtn |= {s for s in dir(builtins) if filt(s, prefix)}
+    rtn |= {s for s in dir(builtins) if filt(s, prefix)}
+    if not rtn:
+        prefix = re.split('\(|=|{|[', prefix)[-1]
```
Comment (laerus) [inline xonsh/completers/python.py]: maybe this should be a raw string?

**Comment by gforsyth** [comment]
I made it a raw string for good measure -- that error was caused by my not escaping the `[` in the split characters.  I just realized, though, that this can lead to infinite recursion.  probably need to fix that before merging

**Comment by laerus** [comment]
yup completion is fixed and i just hit the inf recursion with 
`print(matplotlib,matp<TAB>`

**Comment by gforsyth** [comment]
ok @laerus -- I fixed the infinite recursion and also added splitting on commas (based on your example above).  let me know if you think there should be any other characters we should split the prefix on

**Comment by laerus** [comment]
cool, ty @gforsyth! I'm not really familiar with the completion mechanism yet, i must go through the codebase at some point.

i was checking `@(` completion as it was discussed a little at  [issue #922 comment](https://github.com/xonsh/xonsh/issues/922#issuecomment-235592919)
and it seems that in this branch i get:

```
echo @(<TAB>
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/laerus/.local/lib/python3.5/site-packages/prompt_toolkit/interface.py", line 83n
    completions = list(buffer.completer.get_completions(document, complete_event))
  File "/home/laerus/projects/xonsh/xonsh/ptk/completer.py", line 38, in get_completions
    self.ctx)
  File "/home/laerus/projects/xonsh/xonsh/completer.py", line 36, in complete
    out = func(prefix, line, begidx, endidx, ctx)
  File "/home/laerus/projects/xonsh/xonsh/completers/python.py", line 77, in complete_python_e
    return set(prefix_start + i for i in python_matches)
  File "/home/laerus/projects/xonsh/xonsh/completers/python.py", line 77, in <genexpr>
    return set(prefix_start + i for i in python_matches)
TypeError: Can't convert 'set' object to str implicitly
```

should we just merge this and open a new PR about that?

**Comment by gforsyth** [comment]
Ok @laerus -- thanks for testing this out thoroughly.  I fixed that latest error, although python mode completions don't seem to work very well.  That is probably for another PR, though.  This is ready for another look.  And I rebased on master so the tests should pass, too.  Yay!

File: xonsh/completers/python.py [hunk 2]
```diff
@@ -47,6 +62,7 @@ def complete_python(prefix, line, start, end, ctx):
             rtn |= attr_complete(prefix, ctx, filt)
         rtn |= {s for s in ctx if filt(s, prefix)}
     rtn |= {s for s in dir(builtins) if filt(s, prefix)}
+    rtn |= {s for s in dir(builtins) if filt(s, prefix)}
```
Comment (laerus) [inline xonsh/completers/python.py]: is this line duplicate on purpose?

Comment (gforsyth) [inline xonsh/completers/python.py]: Oops.  Thanks for catching that!

**Comment by laerus** [comment]
thank you @gforsyth! merging this.

**Pull request merged**